File: longest_substring/longest_substring.py
#!/usr/bin/env python3

def longest_substring(s):
    tracker = set()
    longest = 0
    left = 0

    for right in range(len(s)):
        while s[right] in tracker:
            tracker.remove(s[left])
            left += 1

        tracker.add(s[right])
        longest = max(longest, right - left + 1)

    return longest


if __name__ == '__main__':
    s = "abcabcbb"
    print(longest_substring(s))

File: longest_substring/longest_substring.go
package main

import (
	"fmt"
)


func longestSubString(s string) int {
	tracker := make(map[rune]bool)
	longest := 0
	left := 0
	runes := []rune(s)

	for right := 0; right < len(runes); right++ {
		for tracker[runes[right]] {
			delete(tracker, runes[left])
			left ++
		}
		tracker[runes[right]] = true
		currentLength := right - left + 1
		if currentLength > longest {
			longest = currentLength
		}
	}
	return longest
}

func main() {
	s := "abcabcbb"
	result := longestSubString(s)
			fmt.Println(result)


}
